eip/EIPS/eip-4337.md

```diff
@@ -58,45 +69,88 @@ The core interface of the entry point contract is as follows:
 
 ```c++
 function handleOps
-    (UserOperation[] calldata ops, address payable beneficiary)
-    public;
+    (UserOperation[] calldata ops, address payable beneficiary
+     AggregatorInfo[] calldata aggregators, uint[] memory execOrder
+    ) public
 
 function simulateValidation
-    (UserOperation calldata userOp)
-    external returns (uint256 preOpGas, uint256 prefund) {
+    (UserOperation calldata userOp, aggregatedSignature)
+    external returns (uint256 preOpGas, uint256 prefund)
+    
+struct AggregatorInfo {
+    IAggregator aggregator;
+    uint count;
+    bytes signature;
+}
 ```
 
 The core interface required for a wallet to have is:
 
-```c++
-function validateUserOp
-    (UserOperation calldata userOp, bytes32 requestId, uint256 missingWalletFunds)
-    external;
+```solidity
+interface IWallet {
+  function validateUserOp
+      (UserOperation calldata userOp, bytes32 requestId, uint256 missingWalletFunds)
+      external;
+}
+```
+The wallet
+* MUST validate the caller is a trusted EntryPoint
+* The requestId is a hash over the userOp (except signature), entryPoint and chainId
+* MUST validate the signature is a valid signature of the `requestId`
+* MUST pay the entryPoint (caller) at least the "missingWalletFunds" (which might be zero, in case current wallet's deposit is high enough)
+* the wallet MAY pay more than this minimum, to cover future transactions
+* Note that in case the request comes through an aggregator, the signature will be empty.
+
+A Wallet that supports aggregated signature should have the interface:
+```solidity
+interface IAggregatedWallet is IWallet {
+  function getAggregator() view returns (address);
+}
+```
+* **getAggregator()** returns the aggregator this wallet supports.
```

Comment:
We had the same discussion of whether we should have `getTrustedForwarder`, `getTrustedForwarders[]` or `isForwarderTrusted` method in the past, and I think the same consideration should be done here. Do you think only allowing one 'aggregator' for a wallet is going to be ok?

Reply:
yes. I do think we need just one.
This is not an arbitrary trusted caller (which you might argue  that EntryPoint is) - it is the single contract that is tied up to your public-key format.
think of it as an extension of the wallet itself, so should be an immutable member
Changing an aggregator probably also entails changing the public key itself. I think that is something you might do as part of "wallet upgrade"

Reply:
And there is no reason for some advanced multisig wallet to support more than one implementation of signatures some time in the future, in theory? Say, a Gnosis Safe who has an ECDSA, BLS and other kinds of co-signers?

Reply:
we define exactly two distinct modes:
1. each wallet handles its own signatures. EntryPoint trust validateUserOp to do all validations.
2. wallet trusts an external entity first to decide what signature should appear in the final userOp, and then do some aggregation magic.
in your case, if you want a wallet that supports arbitrary number of verification methods, you'd create a single aggregator to manage it.
Technically, you can even have ECDSA signature managed by an aggregator, and on-chain it is not much less efficient (one more call to aggregator.validateSignatures() to validate all ECDSA signatures..)